File: cronista/client/banner.py
"""Banner de abertura "cronista" (docs/17-identidade-visual-cli.md §7).
Degradê por CARACTERE numa grade 2D (linha e coluna), não por letra.
"""

from __future__ import annotations

import pyfiglet
from rich.console import Console
from rich.text import Text

from cronista.client.colors import lerp_color
from cronista.client.signal_bar import VOCE_COLOR

_WORD = "CRONISTA"
_FONT = "ansi_shadow"

_HIGHLIGHT = "#FFFFFF"
_TOP = "#F7ECDA"
_ACCENT = VOCE_COLOR  # #DFB878
_BASE = "#C9A968"
_SHADE = "#B8934F"  # ponto mais escuro do degradê -- ainda dourado, nunca marrom apagado


def _sheen(t: float) -> str:
    if t < 0.15:
        return lerp_color(_HIGHLIGHT, _TOP, t / 0.15)
    if t < 0.4:
        return lerp_color(_TOP, _ACCENT, (t - 0.15) / 0.25)
    if t < 0.7:
        return lerp_color(_ACCENT, _BASE, (t - 0.4) / 0.3)
    return lerp_color(_BASE, _SHADE, (t - 0.7) / 0.3)


def _rows() -> list[str]:
    return [r for r in pyfiglet.figlet_format(_WORD, font=_FONT).splitlines() if r.strip()]


def width() -> int:
    return max((len(r) for r in _rows()), default=0)


def _big() -> Text:
    rows = _rows()
    n_rows = len(rows)
    n_cols = max(len(r) for r in rows)

    text = Text()
    for row_idx, row in enumerate(rows):
        t_y = row_idx / max(1, n_rows - 1)
        for col_idx, ch in enumerate(row):
            if ch == " ":
                text.append(" ")
                continue
            t_x = col_idx / max(1, n_cols - 1)
            text.append(ch, style=f"bold {_sheen((t_x + t_y) / 2)}")
        text.append("\n")
    return text


def render(console: Console) -> Text:
    """Banner grande se couber na largura do terminal; senão, cai pro
    nome simples colorido. Sem cor se a saída não for terminal (docs/17 §5)."""
    if not console.is_terminal:
        return Text(_WORD.lower())
    if console.width >= width() + 2:
        return _big()
    return Text(_WORD.lower(), style=f"bold {VOCE_COLOR}")

File: cronista-tui/src/Banner.js
import React from "react";
import { Box, Text } from "ink";
import figlet from "figlet";
import { lerpColor } from "./shimmer.js";
import { DOURADO } from "./theme.js";

const h = React.createElement;

// Degradê por CARACTERE numa grade 2D (linha e coluna combinadas), não
// por letra -- dá o efeito de luz vindo de um canto.
const FONT = "ANSI Compact";
const HIGHLIGHT = "#FFFFFF";
const TOP = "#F7ECDA";
const ACCENT = DOURADO; // #DFB878
const BASE = "#C9A968";
const SHADE = "#B8934F"; // ponto mais escuro do degradê -- ainda dourado, nunca marrom apagado

function sheen(t) {
  if (t < 0.15) return lerpColor(HIGHLIGHT, TOP, t / 0.15);
  if (t < 0.4) return lerpColor(TOP, ACCENT, (t - 0.15) / 0.25);
  if (t < 0.7) return lerpColor(ACCENT, BASE, (t - 0.4) / 0.3);
  return lerpColor(BASE, SHADE, (t - 0.7) / 0.3);
}

export function bannerRows(word) {
  return figlet.textSync(word, { font: FONT }).split("\n").filter((row) => row.trim() !== "");
}

export function bannerWidth(word) {
  return Math.max(...bannerRows(word).map((row) => Array.from(row).length));
}

// Estático -- sem sweep/shimmer, isso é só do cabeçalho pequeno.
export default function Banner({ word = "cronista" }) {
  const rows = bannerRows(word);
  const nRows = rows.length;
  const nCols = Math.max(...rows.map((row) => Array.from(row).length));

  return h(
    Box,
    // width: nCols é obrigatório nas duas pontas (aqui e por linha,
    // embaixo) -- espaço em branco no FIM de uma linha, sem nada depois,
    // some da largura que o Ink mede sozinho, e as linhas desalinhariam.
    { flexDirection: "column", width: nCols },
    ...rows.map((row, rowIdx) => {
      const tY = rowIdx / Math.max(1, nRows - 1);
      const chars = Array.from(row);
      return h(
        Box,
        { key: rowIdx, width: nCols },
        ...chars.map((ch, colIdx) => {
          if (ch === " ") return h(Text, { key: colIdx }, " ");
          const tX = colIdx / Math.max(1, nCols - 1);
          return h(Text, { key: colIdx, bold: true, color: sheen((tX + tY) / 2) }, ch);
        })
      );
    })
  );
}
